src/tir/schedule/primitive.h

```diff
@@ -19,12 +19,26 @@
 #ifndef TVM_TIR_SCHEDULE_PRIMITIVE_H_
 #define TVM_TIR_SCHEDULE_PRIMITIVE_H_
 
+#include <tvm/support/random_engine.h>
 #include <tvm/tir/schedule/state.h>
 
 namespace tvm {
 namespace tir {
 
 /******** Schedule: Sampling ********/
+/*!
+ * \brief Sample once category from candidates according to the probability weights.
+ * \param self The schedule to update
+ * \param rand_state The pointer to schedule's random state
+ * \param candidates The candidates
+ * \param probs The probability distribution of the candidates
+ * \param decision The sampling decision, if any
+ * \return The random variable sampled from candidates
+ */
+TVM_DLL int64_t SampleCategorical(support::LinearCongruentialEngine::TRandState* rand_state,
+                                  const Array<Integer>& candidates, const Array<FloatImm>& probs,
+                                  Optional<Integer>* decision);
```

Comment:
I'm a little confused as to why this `SampleCategorical` modifies decision while all the others do not. Could you explain?

Reply:
All the sampling primitives in meta schedule modifies the `decision` passed in to reflect which decision is being made inside the method. This includes:
- SamplePerfectTile
- SampleCategorical
- SampleComputeLocation

Without knowing what the decisions are, we are unable to reproduce a trace with randomness.